### ISSUE TITLE
Excessive RAM usage when syncing large folders

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#4976]
When syncing we build a queue of the files to sync, then sort it according to the desired criteria (size, time, random, etc.). For millions of files this can use a lot of RAM. In some cases we don't really care too much about the exact order and it would be better to use less RAM. For this, we could have a limit on the queue size. When doing a pull and iterating over the database / need list, we stop iterating once the limit is reached. The pull proceeds as normal. When complete, we do another pass and grab more needed files, repeat until done.

**Comment by imsodin** [comment]
I think this will come with a slight disadvantage in terms of renames?

I have a almost finished (pre-polish) branch where I implement disk-spilling slices and maps to restrict RAM usage (for both scanning and pulling). I already started working on it before you posted on the forum, that you consider that overkill (from memory, maybe it was something else). Would such an option be considered, i.e. should I invest some more time to roughly finish it for review?

**Comment by calmh** [comment]
Yeah it would make us miss renames if there are a lot of changes and the split happens between the files. I guess that would happen if a large directory is renamed for example.

I guess disk spilling is fine too - but there's a lot of special case sorting etc going on for the pulling. I guess we can implement sort.Interface on top of them. Are they FileInfo (FileIntf?) specific or is there a lot of type assertion going on?

**Comment by calmh** [comment]
I have the implementation for this done and it's small :)

**Comment by imsodin** [comment]
Ok, then probably don't delay for my stuff, because merging till next Tuesday is hardly realistic.

I probably shouldn't start describing it too much, that would just cause confusion. The caller needs to pass a type implementing an interface defined by the disk spilling code. I pushed the code if you want to have a quick look at it, interfaces should be final. However please do not judge that code (nor commit history), it was never meant to be public  in this form x)
`Value`:  https://github.com/syncthing/syncthing/compare/master...imsodin:diskOverflow-nextgen#diff-8d3f8add196bc2edef7d73cdf08e37daR41
`SortValue`: https://github.com/syncthing/syncthing/compare/master...imsodin:diskOverflow-nextgen#diff-c7f82e327453e4ad6fde542f76465758R59

**Comment by AudriusButkevicius** [comment]
Shows nothing to compare. anyways, the crux of the rename problem comes from scans, and we should fix it there. Disk spilling thing will solve that hopefully, and then we can augment the puller to support that.

**Comment by imsodin** [comment]
I updated the links (and wrote to github support, something is broken here). Anyway, the takeaway for me is that a disk-overflow type is probably good to have mid to long term, but this issue can be solved quickly without it - i.e. it's worth cleaning up for preliminary review, but no rush.

**Comment by calmh** [comment]
@imsodin was looking at your branch but there's no definition for `Value`, I think checking is missing a file or so?

**Comment by imsodin** [comment]
I indeed forgot to `git add` a crucial file - symptom of hardly writing any original code at all :)  
I cleaned it up and filed a (wip) PR, referencing a branch is too cumbersome: #4980

**Comment by icolwell** [comment]
Did this fix make it into syncthing yet? Definitely would love to use this feature as I am currently experiencing excessive RAM usage and don't know how else to limit the RAM.

**Comment by sifourquier** [comment]
Hi
I think i have a bug due to this issue

I syncronise the source of QT and syncthing use 13G ram (i have 16G ram)

**Comment by wiwie** [comment]
I may also be affected by this issue. I'm syncing a folder of 114GB, containing about 8k files. Memory usage on a NAS is maxing out at 1GB. I feel this is definitely too much for 8k files. Has this been merged into master? Is there a flag I need to set somewhere, to limit the queue size?

**Comment by jeffbl** [comment]
I also don't know if this is my issue, but I'm also having RAM issues, with very large temporary spikes. On a 4GB Raspberry pi 4 running nothing but syncthing and a tmux remote console session, it is crashing:

runtime: out of memory: cannot allocate 40370176-byte block (2855698432 in use)                                                         
fatal error: out of memory

Most of the time when I check it is using less than a GB of RAM. I have the order for all folders on this device set to "random" so I would expect it not to do any sorting, but the first post above implies that it actually does a sort when "random" is selected? Can it just skip the sort on "random", if this could be contributing to this issue?

**Comment by Leicas** [comment]
I think I have the same issue, on my machine Syncthing used more than 4GB of RAM and made the full system unresponsive. I was syncing a folder of 200GB with mainly 4GB files (50files total). I have other shares but there was no activity on them at the time, and usage is back to 500MB now.

**Comment by uok** [comment]
@jeffbl, @Leicas please post your system/OS specs. 
Some users have seen improvements after setting database tuning in advanced setting to `small`

**Comment by Leicas** [comment]
Same issues on all my machines, Rasperry PI 3B+ on Raspbian Buster, Custom NAS on Debian Stretch, Laptop on Windows 10 (+5GB of RAM used...)...

Changing the database tuning setting to small did solve the issue, just have to do that on all the nodes...

**Comment by AudriusButkevicius** [comment]
Are they all running 32 bit builds?

**Comment by scienmind** [comment]
I've noticed this behavior as well, on a 64-bit Windows machine with 8GB RAM, with fairly large share (~3TB). 
Adding few big files to it, will cause ST's RAM usage to spike up to 2-3GB. 

Once the remote devices are synced up, the RAM settles down back to around 400MB.

**Comment by jeffbl** [comment]
Yes, standard Raspbian on pi4 4GB RAM, which I believe is still 32-bit. I also put in the "small" flag. However, I also noticed that the index-v0.14.0.db folder had 3,949,887 in it, whereas my main server with has a superset of that syncthing data has only 30 files in it. I did a "syncthing -reset-database", which took all night to clear it out. I fired it back up (with the "small") change as well, and it rebuilt to contain about 900 files in the database directory, and is currently running without crashing.

**Comment by imsodin** [comment]
Can anyone that experiences this on a 64bit system please grab heap profile during hihg memory usage (https://docs.syncthing.net/users/profiling.html). And ideally also after it went back to normal for a baseline.

**Comment by scienmind** [comment]
So I reproduced this again (on Win 64 system) and grabbed a bunch of pprof reports over the course of the test: before the addition of the files, during the high ram usage and after. 

Surprisingly enough, the heap profile doesn't look very consistent with the system-reported ram usage:
Once a considerable amount of new files were added, and the syncing to other nodes started, the ram usage raised slightly above 1GB (from an idle 400MB), and kept rising up to 3GB with slight fluctuations, but not getting below 2GB until most of the nodes got in sync, and eventually settling back on 400+MB

However, the pprof's total usage seem to fluctuate in the 100MB-1.7GB range all the time.
But I'm not too familiar with go's profiling system, so I may very much be interpreting it the wrong way.

Long story short, here are the profile files: https://send.firefox.com/download/9c48557ec2e31f90/#7ex2f8MEz81qqDjpHOtP5A

I hope it helps.

**Comment by imsodin** [comment]
@scienmind Most memory usage is due to requests (other devices pulling the data from that device). By default every device can request up to 64MB of data. The largest profile I found (didn't check all of them) reported ~927MB used by the protocol cache. So either you have ~15 syncing connections or more (assuming twice as much memory used as data requested), or you have modified the `maxRequestBytes` setting or there's a bug regarding limiting/caching of requests.

**Comment by scienmind** [comment]
Thanks for taking a look :) 
The largest profile is actually 1719MB (you can see it here: https://send.firefox.com/download/37b2e1718f62d704/#UEgbiOsfJSOi8jdeOnWFOw).

As for connections - the ST instance profiled is actually pretty light one config-wise: while the shared folder itself is big, it is the only share on the device, and it is shared with 3 other devices, that's it.

I haven't edited `maxRequestBytes` manually or any other advanced config value.

**Comment by icolwell** [comment]
@AudriusButkevicius, I was running 64-bit ubuntu 16.04 at the time.

**Comment by Leicas** [comment]
I'm actually still having the issue with the database tuning set to ```small```.
It's happening every time I'm syncing a bunch of ~4GB files on my local 1Gb/s network.
AFAIK it wasn't happening before version 1.3.0 and I was already syncing the same folder.

**Comment by imsodin** [comment]
I completely forgot about the comments in this issue. I am fairly sure it is completely unrelated to the original issue in this thread, but a problem with our pooling in the protocol package, which was not noticeable until a change in go1.13 (or our package is fine, and there is a regression in go1.13, but I'd expect it's our package).

Please do not respond to this issue about the recent memory usage problem, but use the following forum thread: https://forum.syncthing.net/t/large-mem-usage-on-sync-regardless-of-db-tuning/14015  
I intend to investigate further in the near future and open a ticket once the problem is located.

**Comment by abs0** [comment]
Pretty sure I've also hit this - in trying to convert a periodic rsync between one primary and two copies to syncthing, and having to stop syncthing due to memory usage (it reached 16.3g VIRT 10.2g RES)

The single folder to sync had:
- 39,978 directories
- 17,897,687 files
- 1.2TB usage

It counts up to 31,999 directories, then the status remains unchanging while the memory usage keeps climbing. I've tried setting caseSensitiveFS true without apparent effect.

I have syncthing in use in a couple of other cases - but with nothing like the number of files the largest having filers including (10.5TB, 377 directories, 21,231 files) and (0.5TB, 3,816 directories, 49,906 files) and with a steady state of ~1GB SIZE and ~280MB RES. Those are under NetBSD, and the new setup is under Linux, but I don't think the OS is likely to be relevant here.

Going to try to break it down into subfolders and then try to boostrap the sync of each in turn to see if that avoids the issue

Thanks

David

**Comment by AudriusButkevicius** [comment]
You might also want to disable progress updates (check the docs how)

**Comment by calmh** [comment]
@abs0 If you see an issue and can reproduce it, take a heap profile (https://docs.syncthing.net/users/profiling.html) and we can look at it. This issue, as originally described at least, is really about the case where we do an initial sync of a gazillion files, which doesn't sound like what you're doing.

**Comment by Avamander** [comment]
I think I've stumbled upon this issue with Syncthing on ZFS scanning a lot of tiny files in one share (~600GB in total). There's about 6GB free memory when Syncthing isn't running on average, but Syncthing gets OOM killed. Probably because ARC cache (which Syncthing starts to inflate) isn't freed quickly enough? Maybe it would be possible to have a special mode to do things slower but avoid using a lot of memory that could lead to an OOM condition?

**Comment by calmh** [comment]
Again, as noted above, a memory profile would help narrow down what you're suffering from.

**Comment by hohmd** [comment]
> 

Any suggestion how to make memory usage lower? Walker is an issue.
[syncthing-heap-linux-arm-v1.23.4-023015.zip](https://github.com/syncthing/syncthing/files/11410795/syncthing-heap-linux-arm-v1.23.4-023015.zip)

**Comment by imsodin** [comment]
@hohmd 
That's scanning not syncing, and there are knobs to reduce/remove that memory usage. However the issue tracker isn't the place for questions/support. Please use the forum - https://forum.syncthing.net/ - for support, first by searching for any similar issues and then by posting a new one when required. The forum has a lot more eyes and friendly people on it than the issue tracker.

Thanks!

**Comment by fmnets** [comment]
I faced the same problem with large folders so the service was freezing at 700MiB docker container limit and abusing storage intensively. After upgrading the syncthing image to make sure it was compiled with go 1.19 I added GOMEMLIMIT environment variable and set desired soft memory limit. After that it works just fine without any noticeable performance impact. Feels like I can freely put much lower values like 128MiB, but I haven't fully tested that yet, since I'm totally fine with 512.
I believe for older syncthing versions compiled with Go prior to 1.19 as a workaround you can experiment with values for GOGC. The lower (try 10-20) the less memory consumption will be but the more frequently garbage collector will be invoked which may hit the performance on too low values.

**Comment by nazar256** [comment]
I faced the same problem with large folders so the service was freezing at 700MiB docker container limit and abusing storage intensively. After upgrading the syncthing image to make sure it was compiled with go 1.19 I added GOMEMLIMIT environment variable and set desired soft memory limit. After that it works just fine without any noticeable performance impact. Feels like I can freely put much lower values like 128MiB, but I haven't fully tested that yet, since I'm totally fine with 512.
I believe for older syncthing versions compiled with Go prior to 1.19 as a workaround you can experiment with values for GOGC. The lower (try 10-20) the less memory consumption will be but the more frequently garbage collector will be invoked which may hit the performance on too low values.

**Comment by gaby** [comment]
We are running into this issue too. We have one syncthing container per NAS, and they start using 20-30GB of RAM each.

Using latest v2 release as of today
